File: backend/schemas/problems.py
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, ConfigDict


class ProblemSummary(BaseModel):
    id: int
    title: str
    content: str
    domain: str
    difficulty: int
    total_score: int
    school_level: Optional[str] = None
    model_config = ConfigDict(from_attributes=True)


class ProblemDetail(BaseModel):
    id: int
    title: str
    content: str
    domain: str
    difficulty: int
    total_score: int
    school_level: Optional[str] = None
    # answer, reference_solution은 학생에게 노출 금지 — 필드 포함하지 않음

    model_config = ConfigDict(from_attributes=True)


class ProblemListResponse(BaseModel):
    problems: list[ProblemSummary]


class ProblemListPagedResponse(BaseModel):
    problems: list[ProblemSummary]
    total: int
    page: int
    page_size: int


# ── 교사용 스키마 ─────────────────────────────────────────────

class RubricStep(BaseModel):
    step: int
    description: str
    score: int


class RubricSchema(BaseModel):
    total_score: int
    steps: list[RubricStep]


class ProblemCreate(BaseModel):
    title: str
    content: str
    answer: str
    reference_solution: str
    rubric: RubricSchema
    domain: str = "수학2"
    school_level: str = ""
    difficulty: int = 2


class ProblemUpdate(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    answer: Optional[str] = None
    reference_solution: Optional[str] = None
    rubric: Optional[RubricSchema] = None
    domain: Optional[str] = None
    school_level: Optional[str] = None
    difficulty: Optional[int] = None


class TeacherProblemItem(BaseModel):
    id: int
    title: str
    content: str
    answer: str
    reference_solution: str
    rubric: dict
    domain: str
    school_level: Optional[str] = None
    difficulty: int
    submission_count: int
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


class TeacherProblemListResponse(BaseModel):
    problems: list[TeacherProblemItem]
    total: int
    page: int
    page_size: int


class ProblemDeleteResponse(BaseModel):
    id: int
    deleted: bool
    soft_delete: bool

File: frontend/src/api/problems.ts
import { apiFetch, teacherHeaders } from "@/api/client";

export interface RubricStep {
  step: number;
  description: string;
  score: number;
}

export interface RubricSchema {
  total_score: number;
  steps: RubricStep[];
}

export interface ProblemCreate {
  title: string;
  content: string;
  answer: string;
  reference_solution: string;
  rubric: RubricSchema;
  domain?: string;
  difficulty?: number;
}

export interface TeacherProblemItem {
  id: number;
  title: string;
  content: string;
  answer: string;
  reference_solution: string;
  rubric: Record<string, unknown>;
  domain: string;
  difficulty: number;
  submission_count: number;
  created_at: string;
}

export async function getProblems(): Promise<TeacherProblemItem[]> {
  const data = await apiFetch<{ problems: TeacherProblemItem[] }>("/problems", undefined, "문제 목록 조회 실패");
  return data.problems;
}

export async function getProblem(id: number): Promise<TeacherProblemItem> {
  return apiFetch(`/problems/${id}`, undefined, "문제 조회 실패");
}

export interface TeacherProblemListResponse {
  problems: TeacherProblemItem[];
  total: number;
  page: number;
  page_size: number;
}

export async function getTeacherProblems(params?: {
  page?: number;
  page_size?: number;
  has_submissions?: boolean;
}): Promise<TeacherProblemListResponse> {
  const query = new URLSearchParams();
  if (params?.page) query.set("page", String(params.page));
  if (params?.page_size) query.set("page_size", String(params.page_size));
  if (params?.has_submissions) query.set("has_submissions", "true");
  const url = `/teacher/problems${query.toString() ? `?${query.toString()}` : ""}`;
  return apiFetch(url, { headers: teacherHeaders() }, "문제 목록 조회 실패");
}

export async function createProblem(data: ProblemCreate): Promise<TeacherProblemItem> {
  return apiFetch("/teacher/problems", {
    method: "POST",
    headers: teacherHeaders(),
    body: JSON.stringify(data),
  }, "문제 생성 실패");
}

export async function updateProblem(id: number, data: Partial<ProblemCreate>): Promise<TeacherProblemItem> {
  return apiFetch(`/teacher/problems/${id}`, {
    method: "PUT",
    headers: teacherHeaders(),
    body: JSON.stringify(data),
  }, "문제 수정 실패");
}

export async function deleteProblem(id: number): Promise<{ id: number; deleted: boolean; soft_delete: boolean }> {
  return apiFetch(`/teacher/problems/${id}`, {
    method: "DELETE",
    headers: teacherHeaders(),
  }, "문제 삭제 실패");
}
